### PR TITLE
[NUI] Make PropertyMap look like Dictionary.

### PR DESCRIPTION
**Pull request by huayongxu**
### Description of Change ###

1) Some APIs are added like Keys, Values, Contains, Remove, etc.
2) Check whether key is duplicated or not when inserting/adding some key value pair.
3) This patch is related to the patch in csharp-binder:
https://review.tizen.org/gerrit/#/c/platform/core/uifw/dali-csharp-binder/+/272447/

### API Changes ###

 - ACR:

### CODE REVIEW

**Comment by myroot** [comment]
I hope to support this use case
https://github.com/Samsung/TizenFX/blob/1f7646a5a21e7e804449fa6cf62599504f7c141d/src/Tizen.NUI/src/public/Common/PropertyValue.cs#L642

**Comment by huayongxu** [comment]
> I hope to support this use case
> 
> https://github.com/Samsung/TizenFX/blob/1f7646a5a21e7e804449fa6cf62599504f7c141d/src/Tizen.NUI/src/public/Common/PropertyValue.cs#L642

Do you mean that some APIs are added into PropertyValue?
This patch is to add some APIs into PropertyMap.
It is for the issue  https://github.sec.samsung.net/dotnet/NUIBackend/issues/80 .

**Comment by myroot** [comment]
PropertyMap contents could be update from Dali layer such as `PropertyValeu.Get` method/
in this case, your code not work propertly

File: src/Tizen.NUI/src/public/Common/PropertyMap.cs [hunk 1]
```diff
@@ -41,13 +44,43 @@ public PropertyMap() : this(Interop.PropertyMap.NewPropertyMap(), true)
...
         public PropertyMap(PropertyMap other) : this(Interop.PropertyMap.NewPropertyMap(PropertyMap.getCPtr(other)), true)
         {
+            if (other != null)
+            {
+                propertyMap = new Dictionary<PropertyKey, PropertyValue>(other.propertyMap);
+            }
             if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
         }
 
         internal PropertyMap(global::System.IntPtr cPtr, bool cMemoryOwn) : base(cPtr, cMemoryOwn)
```
Comment (myroot) [inline src/Tizen.NUI/src/public/Common/PropertyMap.cs:54]: `propertyMap` not created when use this constructor `internal PropertyMap(global::System.IntPtr cPtr, bool cMemoryOwn) : base(cPtr, cMemoryOwn)`

Comment (huayongxu) [reply src/Tizen.NUI/src/public/Common/PropertyMap.cs]: Done.
propertyMap has been removed.

**Comment by huayongxu** [comment]
> PropertyMap contents could be update from Dali layer such as `PropertyValeu.Get` method/ in this case, your code not work propertly

Done.
propertyMap has been removed, key-value pairs would not be cached.

**Review by zhouleonlei — APPROVED** [review]
lgtm

**Comment by xuelian-bai** [comment]
looks good to me

**Review by elishateng — APPROVED** [review]
LGTM

File: src/Tizen.NUI/src/public/Common/PropertyMap.cs [hunk 2]
```diff
@@ -48,6 +48,42 @@ internal PropertyMap(global::System.IntPtr cPtr, bool cMemoryOwn) : base(cPtr, c
         {
         }
 
+        /// <summary>
+        /// Retrieves all keys.
+        /// </summary>
+        /// <returns>A collection of keys.</returns>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public List<PropertyKey> Keys
```
Comment (myroot) [inline src/Tizen.NUI/src/public/Common/PropertyMap.cs:56]:
```suggestion
        public IList<PropertyKey> Keys
```

File: src/Tizen.NUI/src/public/Common/PropertyMap.cs [hunk 3]
```diff
@@ -48,6 +48,42 @@ internal PropertyMap(global::System.IntPtr cPtr, bool cMemoryOwn) : base(cPtr, c
...
+                return keys;
+            }
+        }
+
+        /// <summary>
+        /// Retrieves all values.
+        /// </summary>
+        /// <returns>A collection of values.</returns>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public List<PropertyValue> Values
```
Comment (myroot) [inline src/Tizen.NUI/src/public/Common/PropertyMap.cs:74]:
```suggestion
        public IList<PropertyValue> Values
```

File: src/Tizen.NUI/src/public/Common/PropertyMap.cs [hunk 4]
```diff
@@ -179,19 +229,72 @@ public PropertyMap Add(KeyValue keyValue)
...
             return this;
         }
 
+        /// <summary>
+        /// Removes the element by the specified key.
+        /// </summary>
+        /// <param name="key">The index key to find.</param>
+        /// <returns>True if the element is removed, false otherwise.</returns>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public bool Remove(PropertyKey key)
```
Comment (myroot) [inline src/Tizen.NUI/src/public/Common/PropertyMap.cs:256]: Use exception for the consistency
```suggestion
        public void Remove(PropertyKey key)
```

**Comment by myroot** [comment]
Are you test done?
I wonder `CSharp_Dali_Property_Map_SetValue_IntKey` this method working well when key was not existed on the map

File: src/Tizen.NUI/src/public/Common/PropertyMap.cs (hunk 2 above)
Comment (huayongxu) [reply src/Tizen.NUI/src/public/Common/PropertyMap.cs]: Done.

File: src/Tizen.NUI/src/public/Common/PropertyMap.cs (hunk 3 above)
Comment (huayongxu) [reply src/Tizen.NUI/src/public/Common/PropertyMap.cs]: Done

File: src/Tizen.NUI/src/public/Common/PropertyMap.cs (hunk 4 above)
Comment (huayongxu) [reply src/Tizen.NUI/src/public/Common/PropertyMap.cs]: Done

**Comment by huayongxu** [comment]
> Are you test done? I wonder `CSharp_Dali_Property_Map_SetValue_IntKey` this method working well when key was not existed on the map

Yes. it works.
In NUI side, SetValue is internal, not public:
            internal set
            {
                using (PropertyKey pKey = new PropertyKey(key))
                {
                    SetValue(pKey, value);
                }
            }

There are 2 solutions:
1) Make set public, make insert/add support duplicated keys.
2) Make set internal, make insert/add not check duplicated keys.

This patch is using solution 2).
Maybe solution 1) is better.

What do you think?

**Comment by myroot** [comment]
I don't understand your means, I don't care to use `SetValue`, it should be private. 

And `PropertyMap.Add` and `Insert` are working like a C# Dictonary, 
Dictonary.Add are throw exception when same key was added

My fisrt question is

You are use `SetValue` in Add method, in this case, Key are not existed on map, 
and I also see native dali code, `CSharp_Dali_Property_Map_SetValue_IntKey` it use PropertyMapp's index operator
and it search a key, so I wonder it is working well when key is not exist on the map

**Comment by huayongxu** [comment]
> I don't understand your means, I don't care to use `SetValue`, it should be private.
> 
> And `PropertyMap.Add` and `Insert` are working like a C# Dictonary, Dictonary.Add are throw exception when same key was added
> 
> My fisrt question is
> 
> You are use `SetValue` in Add method, in this case, Key are not existed on map, and I also see native dali code, `CSharp_Dali_Property_Map_SetValue_IntKey` it use PropertyMapp's index operator and it search a key, so I wonder it is working well when key is not exist on the map

Yes, CSharp_Dali_Property_Map_SetValue_IntKey works well if the key does not exist.

File: src/Tizen.NUI/src/public/Common/PropertyMap.cs (hunk 4 above)
Comment (huayongxu) [reply src/Tizen.NUI/src/public/Common/PropertyMap.cs:262]: Dictionary.Remove returns false if key does not exist.
So PropertyMap follows this rule.

**Review by myroot — APPROVED** [review]

**Review by zg2nets — APPROVED** [review]
looks good！

**Pull request merged**